# Proyek Analisis Data: Indek Air Quality
- Nama: Teduh Afriyoko
- Email: teduhafr@gmail.com
- Id Dicoding:teduhafr

## Menentukan Pertanyaan Bisnis
- Bagaimana kualitas udara PM2.5 dan PM10 di kota-kota selama tahun 2013 - 2017? 
- Kapankah kualitas udara (PM2.5 dan PM10) terbersih dan terkotor selama tahun 2013 - 2017?
- Kota mana yang mendapat kualitas udara terbersih selama tahun 2013 - 2017?

### Data Wragling
#### Gathering Data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import glob
import os

In [2]:
#karena datanya terpisah-pisah, sebelum diolah baiknya kita gabung dulu
data_df= pd.concat(map(pd.read_csv, glob.glob(os.path.join('./air_quality/', "*.csv"))))

In [3]:
data_df.head()

No  year  month  day  hour  PM2.5  PM10   SO2   NO2     CO    O3  TEMP  \
0   1  2013      3    1     0    4.0   4.0   4.0   7.0  300.0  77.0  -0.7   
1   2  2013      3    1     1    8.0   8.0   4.0   7.0  300.0  77.0  -1.1   
2   3  2013      3    1     2    7.0   7.0   5.0  10.0  300.0  73.0  -1.1   
3   4  2013      3    1     3    6.0   6.0  11.0  11.0  300.0  72.0  -1.4   
4   5  2013      3    1     4    3.0   3.0  12.0  12.0  300.0  72.0  -2.0   

     PRES  DEWP  RAIN   wd  WSPM       station  
0  1023.0 -18.8   0.0  NNW   4.4  Aotizhongxin  
1  1023.2 -18.2   0.0    N   4.7  Aotizhongxin  
2  1023.5 -18.2   0.0  NNW   5.6  Aotizhongxin  
3  1024.5 -19.4   0.0   NW   3.1  Aotizhongxin  
4  1025.2 -19.5   0.0    N   2.0  Aotizhongxin

pertanyaan. 
bagaimana kualitas udara PM2.5 dan PM10 di kota-kota cina selama tahun 2013 - 2017
Kapankah kualitas udara (PM2.5 dan PM10) terbersih dan terkotor?
kaitkan dengan data di https://www.bmkg.go.id/kualitas-udara/informasi-partikulat-pm25.bmkg
https://www.bmkg.go.id/kualitas-udara/informasi-partikulat-pm10.bmkg

In [6]:
#### Assessing Data

In [4]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 420768 entries, 0 to 35063
Data columns (total 18 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   No       420768 non-null  int64  
 1   year     420768 non-null  int64  
 2   month    420768 non-null  int64  
 3   day      420768 non-null  int64  
 4   hour     420768 non-null  int64  
 5   PM2.5    412029 non-null  float64
 6   PM10     414319 non-null  float64
 7   SO2      411747 non-null  float64
 8   NO2      408652 non-null  float64
 9   CO       400067 non-null  float64
 10  O3       407491 non-null  float64
 11  TEMP     420370 non-null  float64
 12  PRES     420375 non-null  float64
 13  DEWP     420365 non-null  float64
 14  RAIN     420378 non-null  float64
 15  wd       418946 non-null  object 
 16  WSPM     420450 non-null  float64
 17  station  420768 non-null  object 
dtypes: float64(11), int64(5), object(2)
memory usage: 61.0+ MB


In [ ]:
#terdapat beberapa data yang tidak sama, karena jumlahnya besar akan dihapus data-data yang kosong.

In [8]:
#mencari data yang duplikat
print("Jumlah duplikasi: ", data_df.duplicated().sum())

Jumlah duplikasi:  0


In [21]:
#lihat data statistiknya
data_df.describe()

No           year          month            day  \
count  420768.000000  420768.000000  420768.000000  420768.000000   
mean    17532.500000    2014.662560       6.522930      15.729637   
std     10122.116943       1.177198       3.448707       8.800102   
min         1.000000    2013.000000       1.000000       1.000000   
25%      8766.750000    2014.000000       4.000000       8.000000   
50%     17532.500000    2015.000000       7.000000      16.000000   
75%     26298.250000    2016.000000      10.000000      23.000000   
max     35064.000000    2017.000000      12.000000      31.000000   

                hour          PM2.5           PM10            SO2  \
count  420768.000000  412029.000000  414319.000000  411747.000000   
mean       11.500000      79.793428     104.602618      15.830835   
std         6.922195      80.822391      91.772426      21.650603   
min         0.000000       2.000000       2.000000       0.285600   
25%         5.750000      20.000000      36.000000       3.000000   
50%        11.500000      55.000000      82.000000       7.000000   
75%        17.250000     111.000000     145.000000      20.000000   
max        23.000000     999.000000     999.000000     500.000000   

                 NO2             CO             O3           TEMP  \
count  408652.000000  400067.000000  407491.000000  420370.000000   
mean       50.638586    1230.766454      57.372271      13.538976   
std        35.127912    1160.182716      56.661607      11.436139   
min         1.026500     100.000000       0.214200     -19.900000   
25%        23.000000     500.000000      11.000000       3.100000   
50%        43.000000     900.000000      45.000000      14.500000   
75%        71.000000    1500.000000      82.000000      23.300000   
max       290.000000   10000.000000    1071.000000      41.600000   

                PRES           DEWP           RAIN           WSPM  
count  420375.000000  420365.000000  420378.000000  420450.000000  
mean     1010.746982       2.490822       0.064476       1.729711  
std        10.474055      13.793847       0.821004       1.246386  
min       982.400000     -43.400000       0.000000       0.000000  
25%      1002.300000      -8.900000       0.000000       0.900000  
50%      1010.400000       3.100000       0.000000       1.400000  
75%      1019.000000      15.100000       0.000000       2.200000  
max      1042.800000      29.100000      72.500000      13.200000

### Cleaning Data


In [22]:
#menghapus data yang kosong
data_df.dropna(axis=0, inplace=True)

In [23]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 382168 entries, 0 to 35063
Data columns (total 18 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   No       382168 non-null  int64  
 1   year     382168 non-null  int64  
 2   month    382168 non-null  int64  
 3   day      382168 non-null  int64  
 4   hour     382168 non-null  int64  
 5   PM2.5    382168 non-null  float64
 6   PM10     382168 non-null  float64
 7   SO2      382168 non-null  float64
 8   NO2      382168 non-null  float64
 9   CO       382168 non-null  float64
 10  O3       382168 non-null  float64
 11  TEMP     382168 non-null  float64
 12  PRES     382168 non-null  float64
 13  DEWP     382168 non-null  float64
 14  RAIN     382168 non-null  float64
 15  wd       382168 non-null  object 
 16  WSPM     382168 non-null  float64
 17  station  382168 non-null  object 
dtypes: float64(11), int64(5), object(2)
memory usage: 55.4+ MB


### Exploratory Data Analysis (EDA)
#### Explore

In [24]:
data_df.describe(include='all')

No           year          month            day  \
count   382168.000000  382168.000000  382168.000000  382168.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean     17955.931070    2014.714905       6.475087      15.711308   
std      10001.787087       1.160266       3.449317       8.803064   
min          1.000000    2013.000000       1.000000       1.000000   
25%       9610.000000    2014.000000       3.000000       8.000000   
50%      18103.000000    2015.000000       6.000000      16.000000   
75%      26515.000000    2016.000000       9.000000      23.000000   
max      35064.000000    2017.000000      12.000000      31.000000   

                 hour          PM2.5           PM10            SO2  \
count   382168.000000  382168.000000  382168.000000  382168.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean        11.575184      79.432383     104.573837      15.634814   
std          6.933552      80.154901      91.379446      21.306103   
min          0.000000       2.000000       2.000000       0.285600   
25%          6.000000      20.000000      36.000000       2.000000   
50%         12.000000      55.000000      82.000000       7.000000   
75%         18.000000     111.000000     145.000000      19.000000   
max         23.000000     844.000000     999.000000     500.000000   

                  NO2             CO             O3           TEMP  \
count   382168.000000  382168.000000  382168.000000  382168.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean        50.570068    1229.940563      57.376676      13.518694   
std         35.062086    1157.151476      56.709013      11.425355   
min          2.000000     100.000000       0.214200     -19.900000   
25%         23.000000     500.000000      10.495800       3.100000   
50%         43.000000     900.000000      45.000000      14.400000   
75%         71.000000    1500.000000      82.000000      23.200000   
max        290.000000   10000.000000    1071.000000      41.600000   

                 PRES           DEWP           RAIN      wd           WSPM  \
count   382168.000000  382168.000000  382168.000000  382168  382168.000000   
unique            NaN            NaN            NaN      16            NaN   
top               NaN            NaN            NaN      NE            NaN   
freq              NaN            NaN            NaN   39319            NaN   
mean      1010.813471       2.417195       0.065030     NaN       1.738031   
std         10.452381      13.798402       0.823901     NaN       1.241152   
min        982.400000     -36.000000       0.000000     NaN       0.000000   
25%       1002.400000      -9.000000       0.000000     NaN       0.900000   
50%       1010.400000       3.000000       0.000000     NaN       1.400000   
75%       1019.000000      15.100000       0.000000     NaN       2.200000   
max       1042.800000      29.100000      72.500000     NaN      13.200000   

             station  
count         382168  
unique            12  
top     Nongzhanguan  
freq           33114  
mean             NaN  
std              NaN  
min              NaN  
25%              NaN  
50%              NaN  
75%              NaN  
max              NaN

In [27]:
#mencari corelation di seluruh field
data_df.corr(numeric_only=True)

No      year     month       day      hour     PM2.5      PM10  \
No     1.000000  0.968620  0.056211  0.016624 -0.003728 -0.026866 -0.049872   
year   0.968620  1.000000 -0.192834 -0.008714 -0.003916 -0.029904 -0.044236   
month  0.056211 -0.192834  1.000000  0.017476 -0.001570  0.013363 -0.021498   
day    0.016624 -0.008714  0.017476  1.000000 -0.002153  0.003457  0.024428   
hour  -0.003728 -0.003916 -0.001570 -0.002153  1.000000  0.013275  0.051421   
PM2.5 -0.026866 -0.029904  0.013363  0.003457  0.013275  1.000000  0.884324   
PM10  -0.049872 -0.044236 -0.021498  0.024428  0.051421  0.884324  1.000000   
SO2   -0.227301 -0.168161 -0.223866 -0.006733  0.032342  0.482209  0.466561   
NO2   -0.042262 -0.052299  0.041530  0.013503 -0.002536  0.670753  0.653577   
CO     0.025281  0.012698  0.049558 -0.017784 -0.030150  0.792041  0.703535   
O3    -0.048242 -0.022092 -0.101052 -0.001101  0.281204 -0.151149 -0.114345   
TEMP  -0.133544 -0.167188  0.146401  0.012243  0.139155 -0.128907 -0.094793   
PRES   0.193543  0.194616 -0.022140  0.022391 -0.038386  0.013717 -0.021942   
DEWP  -0.110269 -0.178617  0.283626  0.024001 -0.012506  0.118009  0.072706   
RAIN   0.005496  0.001162  0.017509 -0.002783  0.012610 -0.015268 -0.026885   
WSPM   0.056503  0.091041 -0.142276 -0.013504  0.137995 -0.275686 -0.185789   

            SO2       NO2        CO        O3      TEMP      PRES      DEWP  \
No    -0.227301 -0.042262  0.025281 -0.048242 -0.133544  0.193543 -0.110269   
year  -0.168161 -0.052299  0.012698 -0.022092 -0.167188  0.194616 -0.178617   
month -0.223866  0.041530  0.049558 -0.101052  0.146401 -0.022140  0.283626   
day   -0.006733  0.013503 -0.017784 -0.001101  0.012243  0.022391  0.024001   
hour   0.032342 -0.002536 -0.030150  0.281204  0.139155 -0.038386 -0.012506   
PM2.5  0.482209  0.670753  0.792041 -0.151149 -0.128907  0.013717  0.118009   
PM10   0.466561  0.653577  0.703535 -0.114345 -0.094793 -0.021942  0.072706   
SO2    1.000000  0.501317  0.535289 -0.167093 -0.320991  0.218994 -0.266446   
NO2    0.501317  1.000000  0.705651 -0.476812 -0.276387  0.172287 -0.028276   
CO     0.535289  0.705651  1.000000 -0.315060 -0.323402  0.183800 -0.054173   
O3    -0.167093 -0.476812 -0.315060  1.000000  0.597098 -0.447437  0.313267   
TEMP  -0.320991 -0.276387 -0.323402  0.597098  1.000000 -0.812456  0.820672   
PRES   0.218994  0.172287  0.183800 -0.447437 -0.812456  1.000000 -0.750758   
DEWP  -0.266446 -0.028276 -0.054173  0.313267  0.820672 -0.750758  1.000000   
RAIN  -0.040618 -0.043946 -0.013302  0.023194  0.037182 -0.060999  0.086424   
WSPM  -0.109259 -0.402566 -0.297760  0.296714  0.026664  0.072249 -0.299137   

           RAIN      WSPM  
No     0.005496  0.056503  
year   0.001162  0.091041  
month  0.017509 -0.142276  
day   -0.002783 -0.013504  
hour   0.012610  0.137995  
PM2.5 -0.015268 -0.275686  
PM10  -0.026885 -0.185789  
SO2   -0.040618 -0.109259  
NO2   -0.043946 -0.402566  
CO    -0.013302 -0.297760  
O3     0.023194  0.296714  
TEMP   0.037182  0.026664  
PRES  -0.060999  0.072249  
DEWP   0.086424 -0.299137  
RAIN   1.000000  0.021866  
WSPM   0.021866  1.000000

- karena pembahsan saya berfokus pada PM2.5 dan PM10, saya akan membahas terkait 2 ini saja
-  Jonathan Sarwono hubungan korelasi adalah sebagai berikut :
- 0 : Tidak ada korelasi
- 0.00 - 0.25 : korelasi sangat lemah
- 0.25 - 0.50 : korelasi cukup
- 0.50 - 0.75 : korleasi kuat
- 0.75 - 0.99 : korelasi sangat kuat
- 1 : korelasi sempurna

##### PM2.5
- bisa kita lihat PM2.5 dan PM10 berkorelasi 0.88, jadi ini sangat kuat
- PM2.5 juga memiliki korelasi  untuk NO2 (kuat) dan CO (sangat kuat), 

##### PM10
- memiliki korelasi yagn sangat kuat dengan PM2.5
- memiliki korelasi kuat untuk NO2 dan CO
